crates/gossip/src/harness.rs

```diff
@@ -395,10 +426,197 @@ impl K2GossipFunctionalTestFactory {
+            let space = space.clone();
+            let mut receiver = sender.subscribe();
+            tokio::spawn(async move {
+                tracing::info!("Starting ops stored space informer");
+                while let Ok(op_ids) = receiver.recv().await {
+                    space.inform_ops_stored(op_ids).await.unwrap();
+                }
+
+                tracing::info!("Ops stored sender closed");
+            });
+        }
+
         K2GossipFunctionalTestHarness {
             gossip: space.gossip().clone(),
             space,
             peer_meta_store: Arc::new(peer_meta_store),
         }
     }
 }
+
+/// A factory for creating [K2GossipMemOpStore] instances.
+#[derive(Debug)]
+pub struct K2GossipMemOpStoreFactory {
+    /// The sender for ops stored events.
+    pub ops_stored: tokio::sync::broadcast::Sender<Vec<StoredOp>>,
+}
+
+impl OpStoreFactory for K2GossipMemOpStoreFactory {
+    fn default_config(&self, _config: &mut Config) -> K2Result<()> {
+        Ok(())
+    }
+
+    fn validate_config(&self, _config: &Config) -> K2Result<()> {
+        Ok(())
+    }
+
+    fn create(
+        &self,
+        _builder: Arc<Builder>,
+        space: SpaceId,
+    ) -> BoxFut<'static, K2Result<DynOpStore>> {
+        let sender = self.ops_stored.clone();
+        Box::pin(async move {
+            let op_store: DynOpStore =
+                Arc::new(K2GossipMemOpStore::create(space, sender));
+            Ok(op_store)
+        })
+    }
+}
+
+/// A K2Gossip specialization of the [Kitsune2MemoryOpStore].
+///
+/// This implementation deliberately breaks the op store contract by storing ops with their
+/// "stored at" time set to the "create at" time of the op. This allows gossip tests to create
+/// historical data that won't immediately be synced by the "what's new" mechanism.
+///
+/// The other behaviour change is that this op store can automatically notify the space when ops
+/// are stored, os the DHT model can be updated immediately.
+#[derive(Debug)]
+pub struct K2GossipMemOpStore {
+    inner: Kitsune2MemoryOpStore,
+    ops_stored: tokio::sync::broadcast::Sender<Vec<StoredOp>>,
+}
+
+impl K2GossipMemOpStore {
+    /// Create a new K2GossipMemOpStore.
+    pub fn create(
+        space_id: SpaceId,
+        ops_stored: tokio::sync::broadcast::Sender<Vec<StoredOp>>,
+    ) -> Self {
+        Self {
+            inner: Kitsune2MemoryOpStore::new(space_id),
+            ops_stored,
+        }
+    }
+}
+
+impl OpStore for K2GossipMemOpStore {
+    fn process_incoming_ops(
+        &self,
+        op_list: Vec<Bytes>,
+    ) -> BoxFut<'_, K2Result<Vec<OpId>>> {
+        Box::pin(async move {
+            let ops_to_add = op_list
+                .iter()
+                .map(|op| -> serde_json::Result<(OpId, MemoryOpRecord)> {
+                    let mut op = MemoryOpRecord::from(op.clone());
+
+                    // Behaviour difference from the inner implementation.
+                    op.stored_at = op.created_at;
```

Comment:
I also needed to be able to create historical data without actually fiddling with the internals too much. This seemed like the lightest touch, to pretend that ops turned up at the point they were created. Almost as though they were authored.

It's not true to the real implementation because then other nodes syncing this data will do the same and the "what's new" won't behave as expected. That's why using this op store in the test hardness is optional.

Reply:
That's a fair point, maybe gossip testing needs a more realistic op store... That would be more work to put in place but would actually be cleaner. Yes, let me have a think and then I'll come look at this. I'd like to finish what I'm in the middle of with a Holochain library right now.